test/test_xla_sharding.py

```diff
@@ -118,6 +118,23 @@ def test_inplace_add_with_sharding(self):
     xt.add_(1)  # inplace update should preserve the sharding
     self.assertEqual(sharding_spec, torch_xla._XLAC._get_xla_sharding_spec(xt))
 
+  def test_shard_hashing(self):
+    xt1 = torch.ones(2, 2).to(xm.xla_device())
+    xt2 = torch.ones(2, 2).to(xm.xla_device())
+
+    # Add sharding to xt1, this should result in the hashes being different for
+    # xt1 and xt2
+    xs.mark_sharding(xt1, self._get_mesh((1, self.n_devices)), (0, 1))
```

Comment:
```
t1 = torch.ones(2, 2).to(xm.xla_device())
t2 = t1 * 2
xs.mark_sharding(xt1, self._get_mesh((1, self.n_devices)), (0, 1))
t3 = t1 * 2
```

I think with the current implemenation, `t1` will give different hash after `mark_sharding` will give different hash, but `t2` and `t3` will point to same IR. I felt like that's wrong? I think `SetShardingSpec` should create new IR with sharding and we should not maintain both `sharding_hash_` and `dag_hash_`

Reply:
https://github.com/pytorch/xla/pull/4287#discussion_r1046625399